src/main/java/models/igp/ospf/OSPFInstance.java

```diff
@@ -124,4 +137,84 @@ private void removePrefix(String prefix, String routerId) {
         prefixes.remove(prefixKey);
     }
+
+    public void getShortestPath(String ingressIp, String egressIp) {
+        String ingressRouterId = ipToRouter.get(ingressIp);
+        String egressRouterId = ipToRouter.get(egressIp);
+        OSPFRouter ingressRouter = routers.get(ingressRouterId);
+        OSPFRouter egressRouter = routers.get(egressRouterId);
+
+        // Get ABRs in ingress area and backbone area
+        // Get ABRs in egress are and backbone area
+        // Compute shortest path from ingressRouter --> ingressABR --> egressABR --> egressRouter
+        Map<String, String> ingressAbrs = new HashMap<>();
+        for (String areaId : ingressRouter.areaIds) {
+            OSPFArea area = subgraphs.get(areaId);
+            for (String routerId : area.nodesToInterfaces.keySet()) {
+                if (backboneAbrs.contains(routerId)) {
+                    ingressAbrs.put(routerId, areaId);
+                }
+            }
+        }
+
+        Map<String, String> egressAbrs = new HashMap<>();
+        for (String areaId : egressRouter.areaIds) {
+            OSPFArea area = subgraphs.get(areaId);
+            for (String routerId : area.nodesToInterfaces.keySet()) {
+                if (backboneAbrs.contains(routerId)) {
+                    egressAbrs.put(routerId, areaId);
+                }
+            }
+        }
+
+        Map<String, OSPFPath> ingressToIngressAbr = new HashMap<>();
+        // ingressRouter --> ingressABR
+        for (String ingressAbr : ingressAbrs.keySet()) {
+            OSPFArea area = subgraphs.get(ingressAbrs.get(ingressAbr));
+            OSPFPath path = area.findShortestPathBetweenRouters(ingressRouterId, ingressAbr);
+            ingressToIngressAbr.put(ingressAbr, path);
+        }
+
+        Map<String, Map<String, OSPFPath>> ingressAbrToEgressAbr = new HashMap<>();
+        // ingressABR --> egressABR
+        for (String ingressAbr: ingressAbrs.keySet()) {
+            ingressAbrToEgressAbr.putIfAbsent(ingressAbr, new HashMap<>());
+            for (String egressAbr : egressAbrs.keySet()) {
+                OSPFPath path = subgraphs.get("0.0.0.0").findShortestPathBetweenRouters(ingressAbr, egressAbr);
+                ingressAbrToEgressAbr.get(ingressAbr).put(egressAbr, path);
+            }
+        }
+
+        Map<String, OSPFPath> egressAbrToEgress = new HashMap<>();
+        // egressABR --> egressRouter
+        for (String egressAbr : egressAbrs.keySet()) {
+            OSPFArea area = subgraphs.get(egressAbrs.get(egressAbr));
+            OSPFPath path = area.findShortestPathBetweenRouters(egressAbr, egressRouterId);
+            egressAbrToEgress.put(egressAbr, path);
+        }
+
+        // Find shortest path from ingressRouter --> egressRouter
+        Pair<String, OSPFPath> minPath = null;
+        for (String possibleIngressAbr : ingressAbrs.keySet()) {
+            for (String possibleEgressAbr : egressAbrs.keySet()) {
+                OSPFPath path = OSPFPath.join(
```

Comment:
can move this into the scope where we know its new min path, so we need to create the OSPFPath then